crates/sui-indexer-alt/src/models/watermarks.rs

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{borrow::Cow, cmp};
+
+use crate::{db::Connection, schema::watermarks};
+use diesel::prelude::*;
+use diesel_async::RunQueryDsl;
+
+#[derive(Insertable, Debug, Clone)]
+#[diesel(table_name = watermarks)]
+pub struct StoredWatermark {
+    pub pipeline: String,
+    pub epoch_hi_inclusive: i64,
+    pub checkpoint_hi_inclusive: i64,
+    pub tx_hi: i64,
+    pub epoch_lo: i64,
+    pub reader_lo: i64,
+    pub timestamp_ms: i64,
+    pub pruner_hi: i64,
+}
+
+/// Fields that the committer is responsible for setting.
+#[derive(AsChangeset, Insertable, Selectable, Queryable, Debug, Clone)]
+#[diesel(table_name = watermarks)]
+pub struct CommitterWatermark<'p> {
+    pub pipeline: Cow<'p, str>,
+    pub epoch_hi_inclusive: i64,
+    pub checkpoint_hi_inclusive: i64,
+    pub tx_hi: i64,
+}
+
+/// Outcomes from extending one watermark with another.
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub enum Ordering {
+    /// The watermark was in the future, so could not be added.
+    Future,
+
+    /// The added watermark was in the past, so the current watermark didn't change.
+    Past,
+
+    /// The added watermark was the successor to the current watermark, so was used in the update.
+    Next,
+}
+
+impl CommitterWatermark<'static> {
+    /// Get the current high watermark for the pipeline.
+    pub async fn get(
+        conn: &mut Connection<'_>,
+        pipeline: &'static str,
+    ) -> QueryResult<Option<Self>> {
+        watermarks::table
+            .select(CommitterWatermark::as_select())
+            .filter(watermarks::pipeline.eq(pipeline))
+            .first(conn)
+            .await
+            .optional()
+    }
+}
+
+impl<'p> CommitterWatermark<'p> {
+    /// A new watermark with the given pipeline name indicating zero progress.
+    pub fn initial(pipeline: Cow<'p, str>) -> Self {
+        CommitterWatermark {
+            pipeline,
+            epoch_hi_inclusive: 0,
+            checkpoint_hi_inclusive: 0,
+            tx_hi: 0,
+        }
+    }
+
+    /// Upsert the high watermark as long as it raises the watermark stored in the database.
+    /// Returns a boolean indicating whether the watermark was actually updated or not.
+    ///
+    /// TODO(amnn): Test this (depends on supporting migrations and tempdb).
+    pub async fn update(&self, conn: &mut Connection<'_>) -> QueryResult<bool> {
+        use diesel::query_dsl::methods::FilterDsl;
+        Ok(diesel::insert_into(watermarks::table)
+            .values(StoredWatermark::from(self.clone()))
```

Comment:
I'm not sure if this is correct. In the `from` implementation, we set some fields of `StoredWatermark` to 0 because, I suppose, the pruner is responsible for updating these fields. But here we are updating all columns of a watermark entry for a pipeline, overwriting what the pruner has written with 0s. I think the correct way is to only update the fields that committer knows about and leave the other fields alone.

Reply:
iiuc the combination of deriving `AsChangeset` on `CommitterWatermark` and then the `.on_conflict().do_update().set(self)` is equivalent to what we have in `pg_indexer_store` with the `.set(excluded...`. the `AsChangeset` trait ignores any fields that are not set, so it should only update the the epoch, cp, and tx.

Reply:
Correct -- this will only fill the rows with all 0's if the row does not exist, and if not, it will use `StoredWatermark` directly, which doesn't include those zeroed out values.